pkg/migration/patches.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2023 Upbound Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migration
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+
+	xpv1 "github.com/crossplane/crossplane/apis/apiextensions/v1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var (
+	regexSlice = regexp.MustCompile(`(.+)\[\d+]`)
+	// this regex captures some malformed expressions
+	// (like mismatches in quotes, etc.)
+	regexMap     = regexp.MustCompile(`(.+)\[["|']?\D+["|']?]`)
+	regexJSONTag = regexp.MustCompile(`([^,]+)?(,.+)?`)
+)
+
+const (
+	jsonTagInlined = ",inline"
+)
+
+func removeInvalidPatches(c runtime.ObjectCreater, gvkSource, gvkTarget schema.GroupVersionKind, patchSets []xpv1.PatchSet, targetTemplate *xpv1.ComposedTemplate) error {
```

Comment:
After reading the code, I think I understand what it's doing. However, even if we are not exporting, we can leave comments that tell roughly what the functions do

Reply:
Thanks @sergenyalcin, this is a good point. Documented each function in the patch-filtering library.